Retry http request when unexpected error occurs

serihiro opened the pull request:
# What will this PR change?
- implement https://github.com/google/google-auth-library-ruby/issues/89
    - retry when request failed or timeout in `GCECredentials#fetch_access_token` and `UserRefreshCredentials#revoke!`
    - when give up retrying, raise `Signet::AuthorizationError` with the string of `original_error.inpsect`

serihiro commented:
@googlebot I signed it!

serihiro commented:
@blowmage 
Could you review this? I implemented for this issue https://github.com/google/google-auth-library-ruby/issues/89 which you wrote.

blowmage commented on lib/googleauth/signet.rb:
```diff
@@ -77,6 +77,22 @@ def notify_refresh_listeners
+
+          if retry_count > 0
+            retry_count -= 1
+            retry
```
Do you want to implement an incremental backoff before the retry?

serihiro replied:
@blowmage 
Thank you for your review and comment!

> Do you want to implement an incremental backoff before the retry?

Yes. 
I think there is no problem. 

I think this rescue block behaves like the following:
- `retry_count == 5` =>retry
- `retry_count == 4` =>retry
- `retry_count == 3` =>retry
- `retry_count == 2` =>retry
- `retry_count == 1` =>retry
- `retry_count == 0` =>raise

example:

```ruby
client = GCECredentials.new
client.retry_with_error do
  puts 'hey'
  raise
end
```

```
hey # first execute
hey # 1st retry
hey # 2nd retry
hey # 3rd retry
hey # 4th retry
hey # 5th retry
# raise Signet::AuthorizationError
```

Do you have something to worry about this behavior?

blowmage replied:
I've usually seen retries with an accompanying backoff. So there is an increasing delay between each retry. That seems to be a common Googleism.

serihiro replied:
I see. So I will fix like this. Is this what you mean?

```ruby
def retry_with_error(max_retry_count = 5)
  retry_count = 0

  begin
    yield
  rescue => e
    raise e if e.is_a?(Signet::AuthorizationError) || e.is_a?(Signet::ParseError)

    if retry_count < max_retry_count
      retry_count += 1
      retry
    else
      msg = "Unexpected error: #{e.inspect}"
      raise(Signet::AuthorizationError, msg)
    end
  end
end
```

blowmage replied:
I would expect to see some sort of increment if delay between the retries.

```ruby
def retry_with_error(max_retry_count = 5)
  retry_count = 0

  begin
    yield
  rescue => e
    raise e if e.is_a?(Signet::AuthorizationError) || e.is_a?(Signet::ParseError)

    if retry_count < max_retry_count
      retry_count += 1
      sleep retry_count*0.3
      retry
    else
      msg = "Unexpected error: #{e.inspect}"
      raise(Signet::AuthorizationError, msg)
    end
  end
end
```

serihiro commented:
@blowmage 
Thank you fore your feedback. I fixed `retry_with_error` method and squashed into 960cac289c398ff2d4527ce8daa6ce1068595ea0 . 
Travis CI job was failed for rubocop error, but I confirmed all specs were passed in my env(ruby 2.3.3p222).

hxiong388 commented:
@serihiro Thanks for submitting this PR.

Our team has recently taken over the ownership of this repository, and I'm in process of triaging the PRs and issues. I'll merge this PR once you can fix the rubocop violations.

hxiong388 commented on lib/googleauth/compute_engine.rb:
```diff
@@ -88,16 +88,19 @@ def on_gce?(options = {})
+          when 404
+            raise(Signet::AuthorizationError, NO_METADATA_SERVER_ERROR)
+          else
+            msg = "Unexpected error code #{resp.status}" + UNEXPECTED_ERROR_SUFFIX
```
This exceeds 80 characters max line length.

hxiong388 commented on lib/googleauth/signet.rb:
```diff
@@ -77,6 +77,25 @@ def notify_refresh_listeners
+        begin
+          yield
+        rescue => e
+          raise e if e.is_a?(Signet::AuthorizationError) || e.is_a?(Signet::ParseError)
```
This exceeds 80 characters max line length.

serihiro commented:
@hxiong388 
I fixed rubocop violations and rebased master branch.
Could you check this again?

hxiong388 commented:
@serihiro Thanks! Looks great.
